client/src/components/FilesDialog/FilesDialog.vue

```diff
@@ -1,28 +1,345 @@
+<script setup lang="ts">
+import { library } from "@fortawesome/fontawesome-svg-core";
+import { faCaretLeft } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
+import { BAlert, BButton } from "bootstrap-vue";
+import Vue, { computed, onMounted, ref } from "vue";
+
+import { UrlTracker } from "@/components/DataDialog/utilities";
+import { isSubPath } from "@/components/FilesDialog/utilities";
+import { selectionStates } from "@/components/SelectionDialog/selectionStates";
+import { useConfig } from "@/composables/config";
+import { errorMessageAsString } from "@/utils/simple-error";
+
+import { DirectoryRecord, FileRecord, Model, RecordItem } from "./model";
+import { browseRemoteFiles, FilesSourcePlugin, getFileSources, RemoteEntry } from "./services";
+
+import DataDialogSearch from "@/components/SelectionDialog/DataDialogSearch.vue";
+import DataDialogTable from "@/components/SelectionDialog/DataDialogTable.vue";
+import SelectionDialog from "@/components/SelectionDialog/SelectionDialog.vue";
+
+library.add(faCaretLeft);
+
+interface FilesDialogProps {
+    multiple?: boolean;
+    mode?: "file" | "directory";
+    requireWritable?: boolean;
+    callback?: (files: any) => void;
+}
+
+const props = withDefaults(defineProps<FilesDialogProps>(), {
+    multiple: false,
+    mode: "file",
+    requireWritable: false,
+    callback: () => {},
+});
+
+const { config, isConfigLoaded } = useConfig();
+
+const model = ref<Model>(new Model({ multiple: props.multiple }));
+
+const allSelected = ref(false);
+const selectedDirectories = ref<DirectoryRecord[]>([]);
+const errorMessage = ref<string>();
+const filter = ref();
+const items = ref<RecordItem[]>([]);
+const modalShow = ref(true);
+const optionsShow = ref(false);
+const undoShow = ref(false);
+const hasValue = ref(false);
+const showTime = ref(true);
+const showDetails = ref(true);
+const isBusy = ref(false);
+const currentDirectory = ref<DirectoryRecord>();
+const showFTPHelper = ref(false);
+const selectAllIcon = ref(selectionStates.unselected);
+const urlTracker = ref(new UrlTracker(""));
+
+const fileMode = computed(() => props.mode == "file");
+
+/** Collects selected datasets in value array **/
+function clicked(record: RecordItem) {
+    // ignore the click during directory mode
+    if (!fileMode.value) {
+        return;
+    }
+    if (record.isLeaf) {
+        // record is file
+        selectLeaf(record as FileRecord);
+    } else {
+        // record is directory
+        const directory = record as DirectoryRecord;
+        // you cannot select entire root directory
+        urlTracker.value.atRoot() ? open(directory) : selectDirectory(directory);
+    }
+    formatRows();
+}
```

Comment:
Since you are narrowing here, you should not cast. Narrowing is safer than casting.

```suggestion
function clicked(record: RecordItem) {
    // ignore the click during directory mode
    if (!fileMode.value) {
        return;
    }
    if (record.isLeaf) {
        // record is file
        selectLeaf(record);
    } else {
        // you cannot select entire root directory
        urlTracker.value.atRoot() ? open(record) : selectDirectory(record);
    }
    formatRows();
}
```

In order to be able to narrow, you need to define the type as follows:

```ts
export interface BaseRecordItem {
    id: string;
    label: string;
    details: string;
    isLeaf: boolean;
    url: string;
    labelTitle: string | undefined;
}

export interface FileRecord extends BaseRecordItem {
    isLeaf: true;
}

export interface DirectoryRecord extends BaseRecordItem {
    isLeaf: false;
}

export type RecordItem = FileRecord | DirectoryRecord;
```

Reply:
Done! thank you!